### ISSUE TITLE
Private messages

### ISSUE THREAD

**Issue by noffle** [cabal-club/commons#8]
This issue is to track the implementation of private messages in cabal.

The idea is to add private messages that enable two individual users (feeds) to communicate with each other such that other participants in the cabal will sync their messages, but not be able to read them. The messages will be encrypted using [private-box](https://github.com/auditdrivencrypto/private-box), which uses [chloride](https://github.com/dominictarr/chloride).

There are a few pieces involved across the cabal stack:

## cabal-core (https://github.com/cabal-club/cabal-core/pull/69)
- [x] Core API for publishing messages to a recipient
- [ ] Kappa View for indexing sent and received private messages
- [ ] Core API for reading private message conversations (via ^ kappa view)

## cabal-client
- [ ] API for publishing messages
- [ ] API for reading messages & bubbling up events when they are received, exposed as (virtual?) channels

## Clients
Expose interface for
- [ ] cabal-cli
- [ ] cabal-desktop

**Comment by noffle** [comment]
A major consideration for this feature is preventing abuse & harassment. Blocked & hidden users should NOT be able to have their messages to you be displayed.

Q: where is the best place to handle this logic? I think it makes sense to still index these messages (in case you un{block,hide} someone), but would this be easier to manage in -core or -client?

**Comment by cblgh** [comment]
@noffle good points!

I have been thinking of experimenting with abuse mitigation in the following way:
* introduce some sense of a "small enough" cabal into e.g. cabal-client. when we're small enough, actions can be implicitly assumed to be very high trust (e.g. # identities < 10)
* for small enough cabals, it is okay to show private messages from anyone in the cabal by default
* make a branch that uses [trustnet](https://github.com/cblgh/trustnet)
* for larger cabals (# identities >= 10), private messages will not be shown from people who aren't trusted (e.g. trusted directly, or indirectly as determined by trustnet)
* when bob private messages alice, perform a small check to see if alice trusts bob directly. if not, display a notice in bob's message to the extent of `alice does not appear to have you in their trusted group, so this message might not appear for alice until trust has been issued for you`

I haven't thought about the details of where to index what, but these are the brush strokes with which I have been considering abuse-laden features such as PMs, images & avatars :)

**Comment by okdistribute** [comment]
> A major consideration for this feature is preventing abuse & harassment. Blocked & hidden users should NOT be able to have their messages to you be displayed.
> 
> Q: where is the best place to handle this logic? I think it makes sense to still index these messages (in case you un{block,hide} someone), but would this be easier to manage in -core or -client?

In which repository is the filtering done for blocking/hiding in the channels? Probably best to keep that consistent?

**Comment by noffle** [comment]
For the cabal-client work, a proposal:

- map PMs from `privateMessages.list()` on cabal-core to virtual channels
- use `/join @cblgh` and `/leave @cblgh` syntax for opening/closing these channels
- someone messaging you auto-opens a virtual channel
- logic happens here to prevent a hidden/blocked person who PMs you from opening a virtual channel

**Comment by cblgh** [comment]
command-wise i was thinkin
* `/w kira` && `/pm kira`, the former giving me cosy mmo vibes :)
* cabal-client can take care of the virtual stuff in the /w command under the hood
* we can use the same disambiguation logic that we have for the moderation views, e.g. if multiple kiras we prompt with "hey there are multiple kiras! which one do you wanna pm?"
  * maybe, if we already have a kira pm going, when we do /w non-disambiguated-style we just do a check to see "do i have a convo with someone called kira already?"
* i was thinking of having pm messages also show up in the chat window you are currently viewing as well; it opens up for abuse, but i think it would be interesting to experiment with? there's something about it that i like (again, mmo vibes :)

**Comment by noffle** [comment]
@cblgh I agree on points 1-3. I'm personally -1 on point 4. I already really dislike the status messages we show in regular channels. >__<

**Comment by cblgh** [comment]
@noffle ahh oh no, well we can skip pushing point 4 then. i just wanna experiment with it on my own branch, maybe making it an opt-in for cabal-cli in its config for people who want it, if the experiment ends up working okay ^^
